### PR TITLE
Review of the wrong branch

### PR DESCRIPTION
**Pull request by bifurcation**
This is on the wrong branch, but a lot of it is still relevant on the specialization branch.

### CODE REVIEW

File: draft-rescorla-tls-ctls.md [hunk 1]
```diff
@@ -66,6 +67,20 @@ as a related protocol or as a compression mechanism. Specifically, it
 can be implemented by a layer between the TLS handshake state
 machine and the record layer. See {{compression-layer}} for more details.
 
+## Protocol Embedding
+
+Most of the simplification in this protocol is focused on the TLS handshake,
+since that is where most of the complexity of TLS resides.  This compact
+handshake protocol should be usable in two different contexts: In an integrated
+protocol with the same application sematics as TLS, or as an independent
```
Comment (ekr) [inline draft-rescorla-tls-ctls.md]: semantics.

File: draft-rescorla-tls-ctls.md [hunk 2]
```diff
@@ -66,6 +67,20 @@ as a related protocol or as a compression mechanism. Specifically, it
...
+Authenticated Key Exchange for use in other protocols (e.g., OSCORE).
+
+In the former case, handshake messages would be encapsulated in MLSPlaintext
```
Comment (ekr) [inline draft-rescorla-tls-ctls.md]: TLSPlaintext

File: draft-rescorla-tls-ctls.md [hunk 3]
```diff
@@ -66,6 +67,20 @@ as a related protocol or as a compression mechanism. Specifically, it
...
+objects, as in TLS.  In the latter case, the sequences of handshake messages
+would be carried directly in the application protocol (as TLSCiphertext
+messages, when encrypted).  We assume that the application protocol will delimit
+messages as needed.
```
Comment (ekr) [inline draft-rescorla-tls-ctls.md]: I don't think I understand this text. Isn't it backwards? Why would they be TLSCiphertext in the second version

File: draft-rescorla-tls-ctls.md [hunk 4]
```diff
@@ -106,6 +121,20 @@ a vector with a top range of a varint is denoted as:
...
+
+When a struct is always carried in a length-delimited field (e.g., an extension
+data body), the last vector in the struct does not need to be length-delimited,
+since it takes up all remaining data in the field.  Such vectors are denoted in
+the following form:
+
+~~~~~
+     opaque foo<*>;
+~~~~~
+
```
Comment (ekr) [inline draft-rescorla-tls-ctls.md]: OK, so we'll remove this?

File: draft-rescorla-tls-ctls.md [hunk 5]
```diff
@@ -346,14 +424,16 @@ We can slim down the Certficate message somewhat.
       } CertificateEntry;
 
       struct {
-          CertificateEntry certificate_list[rest of extension];
+          CertificateEntry certificate_list<1..V>;
       } Certificate;
 ~~~~
```
Comment (ekr) [inline draft-rescorla-tls-ctls.md]: Is this just "minimizing unnecessary changes"

File: draft-rescorla-tls-ctls.md [hunk 6]
```diff
@@ -362,6 +442,10 @@ and so may have real problems.
 
 [[OPEN ISSUE: Do we want this at all?]]
 
+[[ XXX-RLB: My opening offer would be to not change this at all.  Just use the
+existing certificate types, but have the compression layer swap the cert_data in
+and out. ]]
+
```
Comment (ekr) [inline draft-rescorla-tls-ctls.md]: I don't think I quite understand how this works.

File: draft-rescorla-tls-ctls.md [hunk 7]
```diff
@@ -395,9 +479,11 @@ Remove the signature algorithm and assume it's tied to the
...
 we just decide to be EC only, it works fine.
 
+[[ XXX-RLB: Unnecessary change.  I would leave it in for now. ]]
+
 ~~~~
       struct {
-          opaque signature[rest of message];
+          opaque signature<1..V>;
       } CertificateVerify;
 ~~~~
```
Comment (ekr) [inline draft-rescorla-tls-ctls.md]: OK.

File: draft-rescorla-tls-ctls.md [hunk 8]
```diff
@@ -407,25 +493,31 @@ Unchanged.
 
 ### HelloRetryRequest
 
-[[TODO]]
+[[ TODO ]]
 
+[[ XXX-RLB: With the one-octet Handshake framing, I think we could probably do
+this in the obvious way. ]]
```
Comment (ekr) [inline draft-rescorla-tls-ctls.md]: OK.

File: draft-rescorla-tls-ctls.md [hunk 9]
```diff
@@ -167,14 +195,12 @@ layer except that the length is a varint.
       } Handshake;
 ~~~~
 
-Overhead: 2 bytes per handshake message (min).
+Overhead: 1 bytes per handshake message.
```
Comment (martinthomson) [inline draft-rescorla-tls-ctls.md]:
```suggestion
Overhead: 1 byte per handshake message.
```

File: draft-rescorla-tls-ctls.md [hunk 10]
```diff
@@ -239,9 +267,42 @@ Overhead: 8 bytes (min)
...
+
+~~~~
+      struct {
+          Random random;
+          CipherSuite cipher_suites<1..V>;
+          Extension extensions<0..V>;
+      } ClientHello;
+~~~~
+
+Overhead: 1 byte (minimum) for the length of the cipher_suites array
```
Comment (martinthomson) [inline draft-rescorla-tls-ctls.md]: 2? the extensions have a length too

File: draft-rescorla-tls-ctls.md (hunk 6 above)
Comment (martinthomson) [reply draft-rescorla-tls-ctls.md]: Define a profile that says:

s/entire Certificate message/a single varint identifier that selects from a table/
